### DISCUSSION TITLE
internal interface `IConstructor`

### DISCUSSION THREAD

**Discussion by viceice** [sebastienros/jint#1141] [Q&A] [answered]
Was this intentional to no longer make that interface public? I use that for a custom type, which implements this interface.

https://github.com/sebastienros/jint/blob/fe4ee89b85a8ea2bd7d3406b55e6597df368b58d/Jint/Native/IConstructor.cs#L5

I'm basicall doing something like this:
https://github.com/sebastienros/jint/blob/fe4ee89b85a8ea2bd7d3406b55e6597df368b58d/Jint.Tests/Runtime/Domain/UuidConstructor.cs#L11

changed in [`v3.0.0-beta-2038`](https://github.com/sebastienros/jint/releases/tag/v3.0.0-beta-2038) via
- #1047

**Comment by viceice** [comment]
I also see PR
-  #1062

So what's the new way to vreate custom types like the `Uuid`? I really need that that's why i send the initial test.

**Comment by lahma** [comment, marked as answer]
Well the idea is to hide as much as possible during while still in beta phase. Not just to irritate people, trying to find how much can be hidden and allow some wiggle room for change. Ideally all test cases should cover public API needs, can you create a PR to show what is needed? 

`IConstructor` would ideally be internal, but maybe we could understand better with an example of usage. Is it basically adding new JS objects that need to be available?

**Reply by viceice** [reply]
i basically use the UUID type, wich is also in your tests. I've created that sample some time ago to fix a similar regression.

so my use case is exactly that UUID type. can we make that type public again? I'll send a PR and add some class comment, so it will be public in future.

not sure how to solve the internal visible to attribute. 🤔 so tests can see all internals

**Reply by viceice** [reply]
- #759

**Reply by viceice** [reply]
maybe there is now a better way to handle / implement such a type?

**Reply by lahma** [reply]
OK I see your point, some things are visible via the `InternalsVisibleTo`, it definitely not the intention to make such extension points hard.

**Reply by viceice** [reply]
So how to go forward?

**Reply by viceice** [reply]
Should we make the test classes public, so it will force the interfaces to be public too?

**Reply by lahma** [reply]
I've created #1144 to track this. Maybe needs a little thought how to make it work. Custom JS object constructors are not a common thing (I guess?), maybe registration via options would work here.
